### ISSUE TITLE
Traits for macro generated Proxies

### ISSUE THREAD

**Issue by DerZade** [dbus2/zbus#433]
It would be really nice to have some traits including the common methods of the generated proxies, so that functions which need any of `new`, `builder`, `into_inner`, `inner` could just use those traits as bounds for generic arguments.

I would imagine something like this:
```rs
trait Proxy<'c> {
    // maybe we could even get away with a default impl here?
    pub async fn new(conn: &::zbus::Connection) -> ::zbus::Result<Self> {
        Self::builder(conn).build().await
    }

    pub fn builder(conn: &::zbus::Connection) -> ::zbus::ProxyBuilder<'c, Self>;

    pub fn into_inner(self) -> ::zbus::Proxy<'c>

    pub fn inner(&self) -> &::zbus::Proxy<'c>
}
```

**Comment by zeenix** [comment]
@DerZade Thanks for filing this. Could you please describe the client side of this with some use case(s)?

**Comment by DerZade** [comment]
Absolutely. I have two utility functions that I want to reuse for multiple proxies.

1. A function that builds a new proxy, provided a path and connection, but also adds some common options when building. Therefore it would need to call the `builder` method of the proxy. It would be nice to just do this via generics, like this:
    ```rs
    async fn build_proxy<T: Proxy<'_>>(conn: &Connection, path: OwnedObjectPath) -> Result<T> {
      let b = T::builder(conn).path(path)?;
      // TODO: Do more stuff with builder
      Ok(builder.build().await)
    }
    ```

1. A function that instantiated a new object with the same path and connection. This therefore needs to extracts the underlying connection and path.
    ```rs
    async fn proxy_with_same_path<I: Proxy<'_>, O: Proxy<'_>>(input: &I) -> Result<O> {
      let p = input.inner();
      O::build_proxy(p.connection(), p.path())
    }
    ```

Both of theses function are in a custom trait and I would love to have a default implementation of that trait that works for all Proxies. Currently I have to implement the trait for all Proxies manually. _(To be 100% correct that is not exactly what I do at the moment, but that does not really matter for this use case. The important takeaway is: I have to implement the same code for multiple proxies, which I could also do in one implementation for a `Proxy` trait)_

That would be my use case, but I bet there are more use cases and functions that would benefit.

**Comment by zeenix** [comment]
@DerZade Thanks so much! That makes a lot of sense IMO. You're in luck because we're in API break and this would require it. If you could provide a PR, that would greatly help us. Otherwise, I'll get back to it at some point. If you do provide a PR, please do read [our contributing guide](https://github.com/dbus2/zbus/blob/main/CONTRIBUTING.md) first.

**Comment by zeenix** [comment]
https://github.com/dbus2/zbus/issues/417 might be relevant here.

**Comment by DerZade** [comment]
Thank you very much for merging this so quickly. Is there already a rough timeline for when the next release is planned?

**Comment by zeenix** [comment]
No, it's an API breaking change so we need to get all those changes done before then. Some of these changes are being worked on by volunteers and I can't say for their priorities and whether I'll have to jump in and finish their PRs for them. 🤷‍♂️

If you'd like to help in accelerating the process, I'd suggest looking into issues tagged with "api-break" label. Do let me know before you take up an issue before starting though to avoid duplication of work.

**Comment by DerZade** [comment]
> No, it's an API breaking change so we need to get all those changes done before then. Some of these changes are being worked on by volunteers and I can't say for their priorities and whether I'll have to jump in and finish their PRs for them. 🤷‍♂️
> 
> If you'd like to help in accelerating the process, I'd suggest looking into issues tagged with "api-break" label. Do let me know before you take up an issue before starting though to avoid duplication of work.

Tbh I don't think this is an breaking change (although I also expected it to be one), since we just added the trait while keeping all the original methods. 
However, I understand that there is no real timeline. 

I'll see if I can find the time to take care of some tickets in the next few days, but I can't promise anything at the moment :)

Thanks a lot for the great work! ❤️

**Comment by zeenix** [comment]
> Tbh I don't think this is an breaking change (although I also expected it to be one), since we just added the trait while keeping all the original methods.

* It is in the sense that it can (at least theoretically) conflict with any existing symbol with the same name. E.g folks have the same idea as you but just not upstreaming it.
* The `main` branch is currently in API break so the next release will be a breaking release in any case. We can roll out a minor release on 3.x branch for sure but I typically only do that for fixes, not new features/API.

> However, I understand that there is no real timeline

:+1: 

> I'll see if I can find the time to take care of some tickets in the next few days, but I can't promise anything at the moment :)

No worries. I was just giving you the option to accelerate the process. Any or all contributions you make are appreciated of course but there is no pressure of any kind. We all have our jobs and life stuff. :)
